libraries/AP_GPS/AP_GPS_GSOF.cpp

```diff
@@ -80,38 +79,48 @@ AP_GPS_GSOF::read(void)
 {
     const uint32_t now = AP_HAL::millis();
 
-    if (gsofmsgreq_index < (sizeof(gsofmsgreq))) {
+    if (gsofmsgreq_index < (gsofmsgreq.count())) {
         const auto com_port = params.com_port.get();
         if (!validate_com_port(com_port)) {
             // The user parameter for COM port is not a valid GSOF port
             return false;
         }
         if (now > gsofmsg_time) {
-            requestGSOF(gsofmsgreq[gsofmsgreq_index], static_cast<HW_Port>(com_port), Output_Rate::FREQ_10_HZ);
+            for (uint16_t i = next_req_gsof; i < gsofmsgreq.size(); i++){
+                if (gsofmsgreq.get(i)) {
+                    next_req_gsof = i;
+                    break;
+                }
+            }
+            requestGSOF(next_req_gsof, static_cast<HW_Port>(com_port), Output_Rate::FREQ_10_HZ);
             gsofmsg_time = now + 110;
             gsofmsgreq_index++;
+            next_req_gsof++;
         }
     }
 
-    bool ret = false;
     while (port->available() > 0) {
         const uint8_t temp = port->read();
 #if AP_GPS_DEBUG_LOGGING_ENABLED
         log_data(&temp, 1);
 #endif
-        const int n_gsof_received = parse(temp, ARRAY_SIZE(gsofmsgreq));
-        if(n_gsof_received == UNEXPECTED_NUM_GSOF_PACKETS) {
-            state.status = AP_GPS::NO_FIX;
-            continue;
-        }
-        const bool got_expected_packets = n_gsof_received == ARRAY_SIZE(gsofmsgreq);
-        ret |= got_expected_packets;
-    }
-    if (ret) {
-        pack_state_data();
+        AP_GSOF::MsgTypes parsed;
+        const int parse_status = parse(temp, parsed);
+        if(parse_status == PARSED_GSOF_DATA) {
+            if (parsed.get(AP_GSOF::POS_TIME) &&
+                parsed.get(AP_GSOF::POS) && 
+                parsed.get(AP_GSOF::VEL) && 
+                parsed.get(AP_GSOF::DOP) && 
+                parsed.get(AP_GSOF::POS_SIGMA)
```

Comment:
This is now robust to extranneous packets being enabled. As long as the data we care about was parsed, we can pack the GPS state structure.